[hudi mor] select timestamp error when jni scan: Failed to open the off-heap table scanner. java exception details: java.io.IOException: Failed to open the hudi MOR slice reader

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
1. create hudi table in spark-sql:
```
 create table hudi_mor_parquet_nocompress(
 uuid int,
 col_boolean boolean,
 col_int int,
 col_long long,
 col_float float,
 col_double double,
 col_decimal decimal(38,18),
 col_date date,
 col_timestamp timestamp,
 col_string string,
 col_binary binary,
 col_array array<int>,
 col_struct struct<a:string,b:int>,
 col_map map<string,int>
 ) USING hudi 
 TBLPROPERTIES(
 type = 'mor',
 primarykey = 'uuid');

 insert into hudi_mor_parquet_nocompress select 1,true,1,1.01,1.001,1.0001,10000000000.0000000001,cast('2020-01-01' as date),cast('2020-01-01 00:00:01' as timestamp),'Top 10 Unsolved Mysteries of Paleontological Dinosaurs, Did You Know?',cast('1110001010101011001001' as binary),array(1,10,100),named_struct('a','first','b',1),map('number-1',1);
 insert into hudi_mor_parquet_nocompress select 2,false,2,2.02,2.002,2.0002,20000000000.0000000002,cast('2020-02-01' as date),cast('2020-02-01 00:00:02' as timestamp),'Xi Jinping, general secretary of the Communist Party of China Central Committee, called on Tuesday for unrelenting efforts in exercising full and rigorous governance over the Party, saying that the CPC will continue to show zero tolerance for corruption.',cast('1110001010101011001001' as binary),array(2,20,200),named_struct('a','second','b',2),map('number-2',2);
 insert into hudi_mor_parquet_nocompress select 3,true,3,3.03,3.003,3.0003,30000000000.0000000003,cast('2020-03-01' as date),cast('2020-03-01 00:00:04' as timestamp),'Nation raises caution on overseas packages',cast('1110001010101011001001' as binary),array(3,30,300),named_struct('a','third','b',3),map('number-3',3);
```
2. sr select:
```
mysql> set hudi_mor_force_jni_reader = false;
Query OK, 0 rows affected (0.00 sec)

mysql> select col_timestamp from hudi_mor_par_parquet_lzo limit 1;
+---------------------+
| col_timestamp       |
+---------------------+
| 2020-10-01 00:00:10 |
+---------------------+
1 row in set (0.07 sec)

mysql> set hudi_mor_force_jni_reader = true;
Query OK, 0 rows affected (0.00 sec)

mysql> select col_timestamp from hudi_mor_par_parquet_lzo limit 1;
ERROR 1064 (HY000): Failed to call the nextChunkOffHeap method of off-heap table scanner. java exception details: java.io.IOException: Failed to get the next off-heap table chunk of hudi.
	at com.starrocks.hudi.reader.HudiSliceScanner.getNext(HudiSliceScanner.java:262)
	at com.starrocks.jni.connector.ConnectorScanner.getNextOffHeapChunk(ConnectorScanner.java:101)
Caused by: java.lang.ClassCastException: class org.apache.hadoop.hive.serde2.io.TimestampWritableV2 cannot be cast to class org.apache.hadoop.io.LongWritable (org.ap

mysql> select col_boolean,col_long,col_float,col_double,col_date,col_struct,col_map from hudi_mor_par_parquet_lzo limit 2;
+-------------+----------+-----------+------------+------------+----------------------+----------------+
| col_boolean | col_long | col_float | col_double | col_date   | col_struct           | col_map        |
+-------------+----------+-----------+------------+------------+----------------------+----------------+
|           1 |        1 |     1.001 |     1.0001 | 2020-01-01 | {"a":"first","b":1}  | {"number-1":1} |
|           0 |        8 |     8.008 |     8.0008 | 2020-08-01 | {"a":"eighth","b":8} | {"number-8":8} |
+-------------+----------+-----------+------------+------------+----------------------+----------------+
2 rows in set (0.35 sec)
```

fe.warn.log:
```
2024-05-31 15:51:27.381+08:00 WARN (starrocks-mysql-nio-pool-14|505) [DefaultCoordinator.getNext():804] query failed: Failed to call the nextChunkOffHeap method of off-heap table scanner. java exception details: java.io.IOException: Failed to get the next off-heap table chunk of hudi.
        at com.starrocks.hudi.reader.HudiSliceScanner.getNext(HudiSliceScanner.java:262)
        at com.starrocks.jni.connector.ConnectorScanner.getNextOffHeapChunk(ConnectorScanner.java:101)
Caused by: java.lang.ClassCastException: class org.apache.hadoop.hive.serde2.io.TimestampWritableV2 cannot be cast to class org.apache.hadoop.io.LongWritable (org.apache.hadoop.hive.serde2.io.TimestampWritableV2 and org.apache.hadoop.io.LongWritable are in unnamed module of loader com.starrocks.utils.loader.ChildFirstClassLoader @75d9aae7)
        at com.starrocks.hudi.reader.HudiColumnValue.getDateTime(HudiColumnValue.java:163)
        at com.starrocks.jni.connector.OffHeapColumnVector.appendValue(OffHeapColumnVector.java:546)
        at com.starrocks.jni.connector.OffHeapTable.appendData(OffHeapTable.java:95)
        at com.starrocks.jni.connector.ConnectorScanner.appendData(ConnectorScanner.java:86)
        at com.starrocks.hudi.reader.HudiSliceScanner.getNext(HudiSliceScanner.java:254)
        ... 1 more
```

### Expected behavior (Required)
select timestamp type success when jni scan

### Real behavior (Required)
select timestamp type error when jni san

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
branch-3.3-05e54ff
```

--- a/java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiColumnValue.java
+++ b/java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiColumnValue.java
@@ -17,6 +17,7 @@ package com.starrocks.hudi.reader;
 import com.starrocks.jni.connector.ColumnType;
 import com.starrocks.jni.connector.ColumnValue;
 import org.apache.hadoop.hive.common.type.HiveDecimal;
+import org.apache.hadoop.hive.serde2.io.TimestampWritableV2;
 import org.apache.hadoop.hive.serde2.objectinspector.ListObjectInspector;
 import org.apache.hadoop.hive.serde2.objectinspector.MapObjectInspector;
 import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
@@ -24,11 +25,15 @@ import org.apache.hadoop.hive.serde2.objectinspector.PrimitiveObjectInspector;
 import org.apache.hadoop.hive.serde2.objectinspector.StructField;
 import org.apache.hadoop.hive.serde2.objectinspector.StructObjectInspector;
 import org.apache.hadoop.hive.serde2.objectinspector.primitive.DateObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.TimestampObjectInspector;
 import org.apache.hadoop.io.LongWritable;
 
 import java.math.BigDecimal;
+import java.sql.Timestamp;
+import java.time.Instant;
 import java.time.LocalDate;
 import java.time.LocalDateTime;
+import java.time.ZoneId;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
@@ -160,9 +165,15 @@ public class HudiColumnValue implements ColumnValue {
 
     @Override
     public LocalDateTime getDateTime(ColumnType.TypeValue type) {
-        Long dateTime = ((LongWritable) fieldData).get();
-        TimeUnit timeUnit = TIMESTAMP_UNIT_MAPPING.get(type);
-        LocalDateTime localDateTime = HudiScannerUtils.getTimestamp(dateTime, timeUnit, timezone);
-        return localDateTime;
+        if (fieldData instanceof Timestamp) {
+            return ((Timestamp) fieldData).toLocalDateTime();
+        } else if (fieldData instanceof TimestampWritableV2) {
+            return LocalDateTime.ofInstant(Instant.ofEpochSecond((((TimestampObjectInspector) fieldInspector)
+                    .getPrimitiveJavaObject(fieldData)).toEpochSecond()), ZoneId.of(timezone));
+        } else {
+            Long dateTime = ((LongWritable) fieldData).get();
+            TimeUnit timeUnit = TIMESTAMP_UNIT_MAPPING.get(type);
+            return HudiScannerUtils.getTimestamp(dateTime, timeUnit, timezone);
+        }
     }
 }

